meshmode/array_context.py

```diff
@@ -437,77 +437,73 @@ def _get_array_container_ac_arith(ary: ArrayContainerWithArithmetic):
 
 # {{{ dataclass containers
 
-# FIXME: Document what this is good for
+# TODO Will this interfere with type-checking, should we want it later?
+# We could always change the definition based on typing.TYPE_CHECKING...
 class NumpyObjectArray(np.ndarray):
-    pass
-
-
-class DataclassArrayContainer(ArrayContainer):
-    """An :class:`ArrayContainer` that implements serialization and
-    deserialization of :func:`~dataclasses.dataclass` fields.
+    """Used as an attribute type with :func:`dataclass_array_container` to
+    document that the attribute will be an object array.
+    :func:`dataclass_array_container` will treat this attribute as an
+    array container.
     """
 
 
-class DataclassArrayContainerWithArithmetic(
-        ArrayContainerWithArithmetic,
-        DataclassArrayContainer):
-    """An :class:`DataclassArrayContainer` where each field is assumed to
-    support arithmetic, e.g. through :class:`ArrayContainerWithArithmetic`.
-    """
+_ARRAY_CONTAINER_LIKE = (ArrayContainer, NumpyObjectArray)
 
 
-@memoize(key=lambda ary: (type(ary), _serialize_dataclass_container))
-def _generate_dataclass_serialization_func(ary):
-    from pytools.codegen import remove_common_indentation
-    code = remove_common_indentation("""
-    def _generated_serialize_dataclass(_ary):
-        return [{}]
-    """.format(", ".join(
-        f"('{f.name}', _ary.{f.name})" for f in fields(ary)
-        if issubclass(f.type, (ArrayContainer, NumpyObjectArray))
-        ))
-    )
+def dataclass_array_container(cls):
```

Comment:
It feels a bit weird to hardcode the arguments to `@dataclass` for this.

Should it take all the arguments from `dataclass` (i.e. `repr`, `init`, etc.)? Or just be a separate decorator that stacks on top of `@dataclass`?

Reply:
6aaf98e